library/objective-c/EnvoyHTTPStreamImpl.m

```diff
@@ -118,47 +122,56 @@ static void ios_on_trailers(envoy_headers trailers, void *context) {
     if (atomic_load(c->canceled) || !callbacks.onTrailers) {
       return;
     }
+
     callbacks.onTrailers(to_ios_headers(trailers));
   });
 }
 
 static void ios_on_complete(void *context) {
   ios_context *c = (ios_context *)context;
   EnvoyHTTPCallbacks *callbacks = c->callbacks;
+  EnvoyHTTPStreamImpl *stream = c->stream;
   dispatch_async(callbacks.dispatchQueue, ^{
-    // TODO: release stream
     if (atomic_load(c->canceled)) {
       return;
     }
+
+    [stream cleanUp];
   });
 }
 
 static void ios_on_cancel(void *context) {
   ios_context *c = (ios_context *)context;
   EnvoyHTTPCallbacks *callbacks = c->callbacks;
-  // TODO: release stream
+  EnvoyHTTPStreamImpl *stream = c->stream;
   dispatch_async(callbacks.dispatchQueue, ^{
     // This call is atomically gated at the call-site and will only happen once.
-    if (!callbacks.onCancel) {
-      return;
+    if (callbacks.onCancel) {
+      callbacks.onCancel();
     }
-    callbacks.onCancel();
+
+    [stream cleanUp];
   });
 }
 
 static void ios_on_error(envoy_error error, void *context) {
   ios_context *c = (ios_context *)context;
   EnvoyHTTPCallbacks *callbacks = c->callbacks;
+  EnvoyHTTPStreamImpl *stream = c->stream;
   dispatch_async(callbacks.dispatchQueue, ^{
-    // TODO: release stream
-    if (atomic_load(c->canceled) || !callbacks.onError) {
+    if (atomic_load(c->canceled)) {
       return;
     }
-    NSString *errorMessage = [[NSString alloc] initWithBytes:error.message.bytes
-                                                      length:error.message.length
-                                                    encoding:NSUTF8StringEncoding];
-    error.message.release(error.message.context);
-    callbacks.onError(error.error_code, errorMessage);
+
+    if (callbacks.onError) {
```

Comment:
Still have the comment about: https://github.com/lyft/envoy-mobile/pull/571#discussion_r341588248. But, it is not introduced by this PR so we can deal with it in a subsequent PR.

Reply:
I'll add the assertions.